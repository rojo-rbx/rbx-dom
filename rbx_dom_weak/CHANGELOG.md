# rbx_dom_weak Changelog

## Unreleased Changes
* Implemented `Default` for `WeakDom`, useful when using Serde or creating an empty `WeakDom`

## 2.6.0 (2023-10-03)
* Added `WeakDom::clone_multiple_into_external` that allows cloning multiple subtrees all at once into a given `WeakDom`, useful for preserving `Ref` properties that point across cloned subtrees ([#364])

[#364]: https://github.com/rojo-rbx/rbx-dom/pull/364

## 2.5.0 (2023-08-09)
* Fix potential stack overflow when creating or inserting into a `WeakDom`. ([#279])
* Added `InstanceBuilder::has_property` for checking if an `InstanceBuilder` defines a given property. ([#283])
* Added `WeakDom::clone_within` and `WeakDom::clone_into_external` for cloning instances into the same or a different `WeakDom`, respectively. ([#312])

[#279]: https://github.com/rojo-rbx/rbx-dom/pull/279
[#283]: https://github.com/rojo-rbx/rbx-dom/pull/283
[#312]: https://github.com/rojo-rbx/rbx-dom/pull/312

## 2.4.0 (2022-06-05)
* Added `WeakDom::into_raw` for enabling fast, non-tree-preserving transformations.
* Added `empty`, `with_class`, and `set_class` methods to `InstanceBuilder`.

## 2.3.0 (2021-10-11)
* Updated to rbx_types 1.3.

## 2.2.0 (2021-07-19)
* Updated to rbx_types 1.2.

## 2.1.0 (2021-07-02)
* Updated to rbx_types 1.1.

## 2.0.0 (2021-06-26) (yanked)
* Updated to rbx_types 1.0.

## 2.0.0-alpha.1 (2021-02-16)
This release is a major, breaking change that introduces many fixes and features.

* `RbxTree` was replaced with `WeakDom`
	* Changed `new` to accept an `InstanceBuilder`, which can contain children.
	* Renamed `get_root_id` to `root_ref`.
	* Renamed `get_instance` to `get_by_ref`.
	* Renamed `get_instance_mut` to `get_by_ref_mut`.
	* Renamed `move_instance` to `transfer`.
	* Renamed `set_parent` to `transfer_within`.
	* Renamed `insert_instance` to `insert`. This method now accepts an `InstanceBuilder`.
	* Removed `remove_instance`, replaced by `destroy` and `transfer`.
	* Removed `iter_all_ids`.
	* Removed `descendants`.
	* Added `root` and `root_mut` for accessing the root instance directly.
* `RbxInstanceProperties` was replaced with `InstanceBuilder`
	* This API is completely different, as it's now a builder. It is now much easier to construct instances.
* `RbxInstance` was replaced with `Instance`
	* This type now directly exposes its values instead of implementing `Deref` for another type.
* Moved types into the `rbx_types` crate, re-exported as `rbx_dom_weak::types`.
	* `RbxId` was replaced with `types::Ref`, which can now represent null referents.
* Added `DomViewer` API from rbx_dom_test to make testing instance trees easier.

## 1.10.1 (2019-12-18)
* Updated `base64`, `md5`, and `uuid` dependencies

## 1.10.0 (2019-09-15)
* Added `RbxTree::set_parent`, for moving instances within a single tree.

## 1.9.0 (2019-07-12)
* Changed `BrickColor` to be much more correct
	* All BrickColor values are now correctly available instead of just palette colors
	* Enum values now align with the `BrickColor.Number` property and `BrickColor.new` constructor
	* `BrickColor` no longer serializes to strings in human-readable Serde formats to avoid information loss. These colors have colliding names:
		* Rust
		* Lilac
		* Gold
		* Deep orange

## 1.8.2 (2019-07-10)
* Added conversion from `Int32` to `BrickColor`. This should fix serialization of `SpawnLocation` instances, which use `int` instead of `BrickColor` in at least the XML model format.

## 1.8.1 (2019-06-20)
* Fixed Serde being able to deserialize `RbxValue` and `RbxValueType`'s internal `__Nonexhaustive` value. This should also fix Serde recommending this value to users in error messages.
* Fixed `BrickColor` serialization to use Roblox color names

## 1.8.0 (2019-06-10)
* Added the `SharedString` type, which is used by Roblox to reduce redundant copies of binary buffers. ([#63](https://github.com/rojo-rbx/rbx-dom/pull/63))

## 1.7.0 (2019-05-14)
* Added conversions from `BrickColor` to `Color3` and `Color3uint8`. ([#46](https://github.com/rojo-rbx/rbx-dom/pull/46))
* Added conversions from `Color3` to `Color3uint8` and vice-versa.
* Changed the human readable serialization of `BinaryString` values to be base64-encoded. This makes JSON-encoded values much smaller. This changes the details of the unspecified serialization of rbx_dom_weak.
* `RbxValueConversion` now derives `PartialEq`.

## 1.6.0 (2019-05-12)
* Added `RbxTree::move_instance` API to move instances from one tree to another.
* Fixed `RbxTree::descendants` to no longer return the instance of the ID you give it. This may break code relying on this (broken) assumption, but was definitely a bug.
* `RbxValueConversion` now derives `Debug` and `Clone` ([#52](https://github.com/rojo-rbx/rbx-dom/issues/52))

## 1.5.0 (2019-05-06)
* Added support for `BrickColor` ([#29](https://github.com/rojo-rbx/rbx-dom/pull/29))
* Added `RbxValue::try_convert_ref`, which can be used to try to coerce values. ([#42](https://github.com/rojo-rbx/rbx-dom/pull/42))
* `Content` values can now be inferred from string literals.
* `RbxValueType` now derives Serde's `Serialize` and `Deserialize` traits

## 1.4.0 (2019-03-27)
* Added support for new types:
	* `NumberRange`
	* [#15](https://github.com/rojo-rbx/rbx-dom/pull/15):
		* `Int64`
		* `Float64`
	* [#20](https://github.com/rojo-rbx/rbx-dom/pull/20):
		* `Rect`
		* `Ray`
		* `ColorSequence`
		* `NumberSequence`

## 1.3.0 (2019-03-14)
* Added `sort_children_by_key` and `sort_children_unstable_by_key` methods to `RbxInstance` to reorder children safely

## 1.2.0 (2019-03-13)
* `RbxTree` and `RbxInstance` are now clonable without an `unimplemented` panic
	* Cloned trees preserve their IDs as-is and there's no public API yet to transplant instances between trees. This is mostly useful for comparing trees before/after a mutation, which we're using in Rojo.

## 1.1.0 (2019-03-11)
* Marked `PhysicalProperties` as `pub`
* Fixed `Serialize` impl for `UnresolvedRbxValue`
	* This removes `Serialize` from `AmbiguousRbxValue`, which should not break any real code.

## 1.0.0 (2019-03-01)
* `RbxValue` and `RbxValueType` can no longer be matched exhaustively, which enables adding new types without breaking code in the future
* Added support for `UDim`, `UDim2`, and `Content` value types
* `PhysicalProperties` is no longer a stub type

## 0.3.1 (2019-02-26)
* Added support for bare bool values when deserializing `UnresolvedRbxValue`
* Implemented `Serialize` for `UnresolvedRbxValue` and `AmbiguousRbxValue`
* Implemented `From<RbxValue>` for `UnresolvedRbxValue`

## 0.3.0 (2019-02-14)
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/rojo-rbx/rbx-dom/pull/8))
* Added `UnresolvedRbxValue` and `AmbiguousRbxValue`, intended to be used alongside `rbx_reflection` to make specifying values less verbose.

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values
