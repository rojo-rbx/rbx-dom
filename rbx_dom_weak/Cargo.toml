[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
version = "2.4.0"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/rojo-rbx/rbx-dom"
repository = "https://github.com/rojo-rbx/rbx-dom.git"
readme = "README.md"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
rbx_types = { version = "1.4.0", path = "../rbx_types", features = ["serde"] }

serde = "1.0.137"

[dev-dependencies]
insta = { version="1.14.1", features=["yaml"] }
