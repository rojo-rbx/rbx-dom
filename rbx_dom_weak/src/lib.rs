//! rbx_dom_weak is a common representation of the Roblox DOM for Rust. It's
//! designed to play nicely with the borrow checker and allows accessing
//! instances by ID in constant time.
//!
//! Constructing a new tree of instances is accomplished by first creating an
//! [`InstanceBuilder`] object that describes a tree of instances and then
//! wrapping it with an [`WeakDom`]:
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let dm = InstanceBuilder::new("DataModel");
//!
//! let mut dom = WeakDom::new(dm);
//!
//! println!("ID of DOM root is {:?}", dom.root_ref());
//! ```
//!
//! Once we have a tree, we can use [`WeakDom::insert`] and
//! [`WeakDom::get_by_ref`] to add instances to the tree and retrieve them.
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let mut dom = WeakDom::new(InstanceBuilder::new("DataModel"));
//!
//! // We can define properties using any type that can be converted to an
//! // rbx_dom_weak::types::Variant.
//! let http_service = InstanceBuilder::new("HttpService")
//!     .with_property("HttpEnabled", true);
//!
//! let http_service_id = dom.insert(dom.root_ref(), http_service);
//!
//! println!("HttpService has ID {:?}", http_service_id);
//! ```
//!
//! To change properties on an instance that's already present in the tree, use
//! [`WeakDom::get_by_ref_mut`]. Note that it isn't possible to add or remove
//! children through this method, use [`WeakDom::insert`] and
//! [`WeakDom::destroy`] instead.

#![deny(missing_docs)]

mod dom;
mod instance;
mod viewer;

pub use rbx_types as types;

pub use ahash::AHashMap;

pub use crate::{
    dom::WeakDom,
    instance::{Instance, InstanceBuilder},
    viewer::{DomViewer, ViewedInstance},
};

/// Helper trait that provides convenience methods for `AHashMap`.
pub trait HashMapExt {
    /// Constructs an empty map.
    fn new() -> Self;

    /// Constructs an empty map with at least the specified capacity.
    fn with_capacity(capacity: usize) -> Self;
}

impl<K, V> HashMapExt for AHashMap<K, V> {
    /// Creates an empty `AHashMap` using the default value for its hasher.
    fn new() -> Self {
        AHashMap::default()
    }

    /// Creates an empty `AHashMap` with at least the specified capacity using
    /// the default value for its hasher.
    fn with_capacity(capacity: usize) -> Self {
        AHashMap::with_capacity_and_hasher(capacity, Default::default())
    }
}
