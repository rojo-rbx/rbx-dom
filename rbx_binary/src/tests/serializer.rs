use rbx_dom_weak::{
    types::{BrickColor, Color3, Color3uint8, Enum, Font, Ref, Region3, SharedString, Vector3},
    InstanceBuilder, WeakDom,
};

use crate::{text_deserializer::DecodedModel, to_writer};

/// A basic test to make sure we can serialize the simplest instance: a Folder.
#[test]
fn just_folder() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder"));
    let mut buffer = Vec::new();

    to_writer(&mut buffer, &tree, &[tree.root_ref()]).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that a tree containing some instances with a value and others
/// without will correctly fall back to (some) default value.
#[test]
fn partially_present() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder").with_children(vec![
        // This instance's `Value` property should be preserved.
        InstanceBuilder::new("StringValue").with_property("Value", "Hello"),
        // This instance's `Value` property should be the empty string.
        InstanceBuilder::new("StringValue"),
    ]));

    let root_refs = tree.root().children();

    let mut buffer = Vec::new();
    to_writer(&mut buffer, &tree, root_refs).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that unknown properties get serialized on instances.
#[test]
fn unknown_property() {
    let tree =
        WeakDom::new(InstanceBuilder::new("Folder").with_property("WILL_NEVER_EXIST", "Hi, mom!"));

    let mut buffer = Vec::new();
    to_writer(&mut buffer, &tree, &[tree.root_ref()]).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that serializing a tree with an unimplemented property type returns
/// an error instead of panicking.
///
/// This test will need to be updated once we implement the type used here.
#[test]
fn unimplemented_type_known_property() {
    let tree = WeakDom::new(InstanceBuilder::new("UIListLayout").with_property(
        "Padding",
        Region3::new(Vector3::new(0.0, 0.0, 50.0), Vector3::new(0.0, 0.0, 50.0)),
    ));

    let mut buffer = Vec::new();
    let result = to_writer(&mut buffer, &tree, &[tree.root_ref()]);

    assert!(result.is_err());
}

/// Ensures that serializing a tree with an unimplemented property type AND an
/// unknown property descriptor returns an error instead of panicking.
///
/// Because rbx_binary has additional logic for falling back to values with no
/// known property descriptor, we should make sure that logic works.
///
/// This test will need to be updated once we implement the type used here.
#[test]
fn unimplemented_type_unknown_property() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder").with_property(
        "WILL_NEVER_EXIST",
        Region3::new(Vector3::new(0.0, 0.0, 50.0), Vector3::new(0.0, 0.0, 50.0)),
    ));

    let mut buffer = Vec::new();
    let result = to_writer(&mut buffer, &tree, &[tree.root_ref()]);

    assert!(result.is_err());
}

/// Ensures that the serializer returns an error instead of panicking if we give
/// it an ID not present in the tree.
#[test]
fn unknown_id() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder"));

    let mut buffer = Vec::new();
    let result = to_writer(&mut buffer, &tree, &[Ref::new()]);

    assert!(result.is_err());
}

#[test]
fn migrated_properties() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder").with_children([
        InstanceBuilder::new("ScreenGui").with_property("ScreenInsets", Enum::from_u32(0)),
        InstanceBuilder::new("ScreenGui").with_property("IgnoreGuiInset", true),
        InstanceBuilder::new("Part").with_property("Color", Color3::new(1.0, 1.0, 1.0)),
        InstanceBuilder::new("Part").with_property("BrickColor", BrickColor::Alder),
        InstanceBuilder::new("Part").with_property("brickColor", BrickColor::Alder),
        InstanceBuilder::new("TextLabel").with_property("FontFace", Font::default()),
        InstanceBuilder::new("TextLabel").with_property("Font", Enum::from_u32(8)),
    ]));

    let mut buffer = Vec::new();

    to_writer(&mut buffer, &tree, &[tree.root_ref()]).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that only one name for each logical property is serialized to a
/// file. Here, we use BasePart.Size and BasePart.size, which alias and both
/// serialize to BasePart.size.
///
/// For fun, we also have a part with no size property at all. It should default
/// to (4.0, 1.2, 2.0), a relic of Roblox's distant past.
#[test]
fn logical_properties_basepart_size() {
    let tree = WeakDom::new(
        InstanceBuilder::new("Folder")
            .with_child(
                InstanceBuilder::new("Part").with_property("Size", Vector3::new(1.0, 2.0, 3.0)),
            )
            .with_child(
                InstanceBuilder::new("Part").with_property("size", Vector3::new(4.0, 5.0, 6.0)),
            )
            .with_child(InstanceBuilder::new("Part")),
    );

    let mut buffer = Vec::new();
    to_writer(&mut buffer, &tree, tree.root().children()).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that all valid combinations of color property names and
/// value types are properly handled.
#[test]
fn part_color() {
    let tree = WeakDom::new(
        InstanceBuilder::new("Folder")
            .with_child(
                InstanceBuilder::new("Part")
                    .with_property("Color3uint8", Color3::new(-0.25, 0.5, 1.2)),
            )
            .with_child(
                InstanceBuilder::new("Part")
                    .with_property("Color3uint8", Color3uint8::new(25, 86, 254)),
            )
            .with_child(
                InstanceBuilder::new("Part").with_property("Color", Color3::new(0.0, 0.5, 1.0)),
            )
            .with_child(
                InstanceBuilder::new("Part").with_property("Color", Color3uint8::new(1, 30, 100)),
            ),
    );

    let mut buf = Vec::new();
    let _ = to_writer(&mut buf, &tree, tree.root().children());

    let decoded = DecodedModel::from_reader(buf.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

#[test]
fn default_shared_string() {
    let mut tree = WeakDom::new(InstanceBuilder::new("Folder"));
    let ref_1 = tree.insert(
        tree.root_ref(),
        InstanceBuilder::new("Model").with_property(
            // This is the first SharedString property I saw in the database
            "ModelMeshData",
            SharedString::new(b"arbitrary string".to_vec()),
        ),
    );
    let ref_2 = tree.insert(tree.root_ref(), InstanceBuilder::new("Model"));

    let mut buf = Vec::new();
    let _ = to_writer(&mut buf, &tree, &[ref_1, ref_2]);

    let decoded = DecodedModel::from_reader(buf.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}
