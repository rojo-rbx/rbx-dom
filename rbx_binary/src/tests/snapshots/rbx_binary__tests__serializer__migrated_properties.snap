---
source: rbx_binary/src/tests/serializer.rs
expression: decoded
---
num_types: 4
num_instances: 8
chunks:
  - Inst:
      type_id: 3
      type_name: Folder
      object_format: 0
      referents:
        - 7
  - Inst:
      type_id: 1
      type_name: Part
      object_format: 0
      referents:
        - 2
        - 3
        - 4
  - Inst:
      type_id: 0
      type_name: ScreenGui
      object_format: 0
      referents:
        - 0
        - 1
  - Inst:
      type_id: 2
      type_name: TextLabel
      object_format: 0
      referents:
        - 5
        - 6
  - Prop:
      type_id: 3
      prop_name: Name
      prop_type: String
      values:
        - Folder
  - Prop:
      type_id: 1
      prop_name: Color3uint8
      prop_type: Color3uint8
      values:
        - - 255
          - 255
          - 255
        - - 180
          - 128
          - 255
        - - 180
          - 128
          - 255
  - Prop:
      type_id: 1
      prop_name: Name
      prop_type: String
      values:
        - Part
        - Part
        - Part
  - Prop:
      type_id: 0
      prop_name: Name
      prop_type: String
      values:
        - ScreenGui
        - ScreenGui
  - Prop:
      type_id: 0
      prop_name: ScreenInsets
      prop_type: Enum
      values:
        - 0
        - 1
  - Prop:
      type_id: 2
      prop_name: FontFace
      prop_type: Font
      values:
        - family: "rbxasset://fonts/families/SourceSansPro.json"
          weight: Regular
          style: Normal
          cachedFaceId: ~
        - family: "rbxasset://fonts/families/Guru.json"
          weight: Regular
          style: Normal
          cachedFaceId: ~
  - Prop:
      type_id: 2
      prop_name: Name
      prop_type: String
      values:
        - TextLabel
        - TextLabel
  - Prnt:
      version: 0
      links:
        - - 0
          - 7
        - - 1
          - 7
        - - 2
          - 7
        - - 3
          - 7
        - - 4
          - 7
        - - 5
          - 7
        - - 6
          - 7
        - - 7
          - -1
  - End
